opm/simulators/linalg/DILU.hpp

```diff
@@ -0,0 +1,183 @@
+    */
+    SeqDilu(const M& A)
+        : A_(A)
+    {
+        Dinv_.resize(A_.N());
+        // the Dinv matrix must be initialised
+        update();
+    }
+
+    /*!
+       \brief Update the preconditioner.
+       \copydoc Preconditioner::update()
+    */
+    virtual void update() override
+    {
+        OPM_TIMEBLOCK(update);
+        
+        auto endi = A_.end();
+        for ( auto row_i = A_.begin(); row_i != endi; ++row_i) {
+            Dinv_[row_i.index()] = A_[row_i.index()][row_i.index()];
+        }
+        
+        for ( auto row_i = A_.begin(); row_i != endi; ++row_i)
+        {
+           auto Dinv_temp = Dinv_[row_i.index()];
+            for (auto a_ij = row_i->begin(); a_ij.index() < row_i.index(); ++a_ij) {
+                auto a_ji = A_[a_ij.index()].find(row_i.index());
+                // if A[i, j] != 0 and A[j, i] != 0
+                if (a_ji != A_[a_ij.index()].end()) {
+                    // Dinv_temp -= A[i, j] * d[j] * A[j, i]
+                    Dinv_temp -= (*a_ij) * Dune::FieldMatrix(Dinv_[a_ij.index()]) * (*a_ji);
+                }
+            }
+            Dinv_temp.invert();
+            Dinv_[row_i.index()] = Dinv_temp;
+        }
+    }
+
+    /*!
+       \brief Prepare the preconditioner.
+       \copydoc Preconditioner::pre(X&,Y&)
+    */
+    virtual void pre(X& v, Y& d) override
+    {
+        DUNE_UNUSED_PARAMETER(v);
+        DUNE_UNUSED_PARAMETER(d);
+    }
+
+
+    /*!
+       \brief Apply the preconditioner.
+       \copydoc Preconditioner::apply(X&,const Y&)
+    */
+    virtual void apply(X& v, const Y& d) override
+    {
+        // M = (D + L_A) D^-1 (D + U_A)   (a LU decomposition of M)
+        // where L_A and U_A are the strictly lower and upper parts of A and M has the properties:
+        // diag(A) = diag(M)
+        // solving the product M^-1(b-Ax) using upper and lower triangular solve
+        // z = x_k+1 - x_k = M^-1(b - Ax) = (D + U_A)^-1 D (D + L_A)^-1 (b - Ax)
+        OPM_TIMEBLOCK(apply);
+        using Xblock = typename X::block_type;
+        using Yblock = typename Y::block_type;
+        
+        // lower triangular solve: (D + L_A) y = b - Ax
+        auto endi = A_.end();
+        for (auto row_i = A_.begin(); row_i != endi; ++row_i)
+        {
+            Yblock rhs = d[row_i.index()];
+            for (auto a_ij = (*row_i).begin(); a_ij.index() < row_i.index(); ++a_ij) {
+                // if  A[i][j] != 0
+                // rhs -= A[i][j]* y[j], where v_j stores y_i
+                a_ij->mmv(v[a_ij.index()], rhs);
+            }
+            // y_i = Dinv_i * rhs
+            // storing y_i in v_i
+            Dinv_[row_i.index()].mv(rhs, v[row_i.index()]);  // (D + L_A)_ii = D_i
+        }
+
+        // upper triangular solve: (D + U_A) z = Dy 
+        auto rendi = A_.beforeBegin();
+        for (auto row_i = A_.beforeEnd(); row_i != rendi; --row_i)
+        {
+            // rhs = 0
+            Xblock rhs;
```

Comment:
Are we 100% sure this will initialize rhs to zero always?

Reply:
If the block is an `OPM::MatrixBlock` this should always be the case:

https://github.com/OPM/opm-models/blob/e945446f48b9bb19ea72f4923f4465e24639625e/opm/simulators/linalg/matrixblock.hh#L236

But I changed it to `Xblock rhs(0.0);` to be on the safe side and increase the readability.